Added DB to backend

farbodahm opened the pull request:
Added Mysql database to backend.
(Make sure to change connection information in `config.json`)

- In this pr, I used [Flask-SQLAlchemy Doc](https://flask-sqlalchemy.palletsprojects.com/en/2.x/) a lot
- Also [this](https://www.youtube.com/watch?v=juPQ04_twtA) video was so helpful.

(Also I should add that `Github CI Test` is going to be failed because we need `mysql server` to pass our integration tests; I'm aware of that and working on it!)

farbodahm committed:
Added first version of DB; Now sign-in is with db

farbodahm committed:
Fully migrated to MySQl

farbodahm committed:
Documentation: Make docs more clear

farbodahm committed:
Updated tests

farbodahm committed:
Updated main doc to use DB & Added new modules to requirements.txt

pesarkhobeee commented:
Dear @hramezani, could you please take a look at this PR? what is the best way to write unit tests and also I will be grateful if you can review the whole PR as it is the fundamental of working with DB in this program :)

hramezani commented on api.py:
```diff
@@ -8,25 +8,47 @@
     CONFIG = json.load(config_file)
 
+# Database connection
+database_username = CONFIG['database']['username']
```
I would suggest to change it like:
```suggestion
database_username = CONFIG.get('database', {}).get('username')
```
Otherwise, You will face `KeyError` if config file doesn't have a `database` key.
Please change the other variables as well.

hramezani commented on common/DbHandler.py:
```diff
@@ -1,73 +1,57 @@
-            user.password.encode('utf-8'), password.encode('utf-8')
-        ):
-            return_message = user
+        # Chech if user exists and password is correct
```
```suggestion
    # Check if user exists and password is correct
```

hramezani commented on common/DbHandler.py:
```diff
@@ -1,73 +1,57 @@
+
+        # Check if id exists and it's owner sends request
+        if link_object:
+            if link_object.owner_id == user_object.id:
```
It is better to include this condition in `link_object` query above.

hramezani commented on common/DbHandler.py:
```diff
@@ -1,73 +1,57 @@
+            if link_object.owner_id == user_object.id:
+                Link.query.filter_by(id=link_id).delete()
+                db.session.commit()
+                return 0
```
It would be good to return proper status instead of `0`, `1`, or `2`.

farbodahm replied:
Would you mind please do me a favor and clarify what do you mean by `proper status`?
Do you mean to use something like [this](https://docs.python.org/3.4/library/os.html#os.EX_OK) or [raising an exception](https://stackoverflow.com/questions/99683/which-and-why-do-you-prefer-exceptions-or-return-codes)?
([As far as I know](https://stackoverflow.com/questions/35818977/does-python-have-an-exit-success-constant), that would be better to throw exceptions instead of returning numbers as status codes; Alongside that throwing exceptions has its own advantages and dis-advantages)

hramezani replied:
You can raise exceptions or use string (like `OK`, `LINK_DOES_NOT_EXISTS`, ...) instead of a number.

hramezani commented on resources/Links.py:
```diff
@@ -10,26 +10,28 @@
+        )
+
+        bookmarked_links_ids = []
+        for link in current_user_object.links:
+            bookmarked_links_ids.append(link.id)
         return make_response(
-            jsonify(bookmarked_links=bookmarked_links_names), 200
+            jsonify(bookmarked_links_id=bookmarked_links_ids), 200
         )
```
```suggestion
        return make_response(
            jsonify(bookmarked_links=bookmarked_links_names), 200
            jsonify(bookmarked_links_id=[link.id for link in current_user_object.links]), 200
        )
```

hramezani commented on resources/Links.py:
```diff
@@ -62,17 +73,26 @@ def post(self):
+        elif remove_status == 2:
             return make_response(
-                jsonify(msg="Link doesn't exists"), 404
+                jsonify(msg="Link doesn't exists!"), 406
```
Why `406`?  :thinking:

hramezani commented on tests/test_user.py:
```diff
@@ -1,26 +1,73 @@
+    """
+    For generating random username
+    """
+    letters = string.ascii_lowercase
```
Remove this variable and use the `string.ascii_lowercase` directly in below line.

farbodahm committed:
Refactor: Changed index operator to get() for lists

farbodahm committed:
Code reaftored

farbodahm committed:
Refactored: Chaned return status from int to str

farbodahm committed:
Doc Updated for return status

farbodahm committed:
Added sqlite db for testing purposes

farbodahm commented:
@hramezani I deeply appreciate your review and your time, it was really helpful, thanks!

pesarkhobeee commented on common/Link.py:
```diff
@@ -1,17 +1,8 @@
+    id = db.Column(db.Integer, primary_key=True, autoincrement=True)
+    url = db.Column(db.String(50))
+    owner_id = db.Column(db.Integer, db.ForeignKey('user.id'))
+    categories = db.Column(db.String(50))
```
In the future, categories should save in their own database table, and here we will have only the foreign keys!

pesarkhobeee commented on README.md:
```diff
@@ -23,6 +23,25 @@ Then you can run
 
+### Database Connection
+
+By default, this code uses `tests/test.db` (which is a sqlite) for testing purposes.
```
this problem will fix with the help of docker in the near future but maybe we should consider adding this choice for users to use either MySQL or SQLite!

hramezani commented on api.py:
```diff
@@ -8,25 +8,49 @@
-api.add_resource(Login, CONFIG['routes']['user']['login'])
-api.add_resource(Signup, CONFIG['routes']['user']['signup'])
-api.add_resource(Links, CONFIG['routes']['user']['links'])
+api.add_resource(Login, CONFIG.get('routes', {}).get('user').get('login'))
```
```suggestion
api.add_resource(Login, CONFIG.get('routes', {}).get('user', {}).get('login'))
```

hramezani commented on common/DbHandler.py:
```diff
@@ -1,73 +1,57 @@
-    def append_new_link(username: str, new_link: Link):
-        USERS[username].append_new_link(new_link)
-        return 0
+    def append_new_link(new_link: Link):
```
It would be great to determine the return type of function and other functions as well:
```suggestion
    def append_new_link(new_link: Link) -> str:
```

hramezani commented on common/User.py:
```diff
@@ -1,22 +1,16 @@
-    def __repr__(self):
-        return f"User(id='{self.id}')"
+    def __init__(self, **kwargs):
+        super(User, self).__init__(**kwargs)
```
Use the Python3 style `super` call:
```suggestion
        super().__init__(**kwargs)
```

hramezani commented on common/Link.py:
```diff
@@ -1,17 +1,8 @@
-        self.categories.append(category)
+class Link(db.Model):
+    id = db.Column(db.Integer, primary_key=True, autoincrement=True)
+    url = db.Column(db.String(50))
```
Is it a `tab` instead of `spaces` at the beginning of line :thinking: ?

farbodahm replied:
Well, I configured my editor to convert `Tab`s to `Space`s for me, it was doing great till now! I'm not sure how that happened!

hramezani reviewed (approved):
LGTM :+1: . I left some small comments.

farbodahm committed:
Added seprate table for categories in db

farbodahm committed:
Changed one-to-many relationship to many-to-many for Link and Category

farbodahm committed:
Dictionary get method return {} if key not found

farbodahm committed:
Added functions return type

farbodahm committed:
Using Python3 style of super() call

farbodahm committed:
Refactored: Added comments and make some cleaning

farbodahm commented:
Thank for your great reviews, I made some small changes based on them; I really appreciate your help.

pesarkhobeee commented:
Dear @farbodahm 
I was testing this pull request, I discovered some error while I was tring to get the links:
```
127.0.0.1 - - [15/Sep/2020 06:44:10] "GET /user/links HTTP/1.1" 500 -
Traceback (most recent call last):
  File "/home/farid/lab/Restfulness-core-api/venv/lib/python3.6/site-packages/flask/app.py", line 2464, in __call__
    return self.wsgi_app(environ, start_response)
  File "/home/farid/lab/Restfulness-core-api/venv/lib/python3.6/site-packages/flask/app.py", line 2450, in wsgi_app
    response = self.handle_exception(e)
  File "/home/farid/lab/Restfulness-core-api/venv/lib/python3.6/site-packages/flask_restful/__init__.py", line 272, in error_router
    return original_handler(e)
  File "/home/farid/lab/Restfulness-core-api/venv/lib/python3.6/site-packages/flask/app.py", line 1867, in handle_exception
    reraise(exc_type, exc_value, tb)
  File "/home/farid/lab/Restfulness-core-api/venv/lib/python3.6/site-packages/flask/_compat.py", line 38, in reraise
    raise value.with_traceback(tb)
  File "/home/farid/lab/Restfulness-core-api/venv/lib/python3.6/site-packages/flask/app.py", line 2447, in wsgi_app
    response = self.full_dispatch_request()
  File "/home/farid/lab/Restfulness-core-api/venv/lib/python3.6/site-packages/flask/app.py", line 1952, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/home/farid/lab/Restfulness-core-api/venv/lib/python3.6/site-packages/flask_restful/__init__.py", line 272, in error_router
    return original_handler(e)
  File "/home/farid/lab/Restfulness-core-api/venv/lib/python3.6/site-packages/flask/app.py", line 1821, in handle_user_exception
    reraise(exc_type, exc_value, tb)
  File "/home/farid/lab/Restfulness-core-api/venv/lib/python3.6/site-packages/flask/_compat.py", line 38, in reraise
    raise value.with_traceback(tb)
  File "/home/farid/lab/Restfulness-core-api/venv/lib/python3.6/site-packages/flask/app.py", line 1950, in full_dispatch_request
    rv = self.dispatch_request()
  File "/home/farid/lab/Restfulness-core-api/venv/lib/python3.6/site-packages/flask/app.py", line 1936, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/home/farid/lab/Restfulness-core-api/venv/lib/python3.6/site-packages/flask_restful/__init__.py", line 468, in wrapper
    resp = resource(*args, **kwargs)
  File "/home/farid/lab/Restfulness-core-api/venv/lib/python3.6/site-packages/flask/views.py", line 89, in view
    return self.dispatch_request(*args, **kwargs)
  File "/home/farid/lab/Restfulness-core-api/venv/lib/python3.6/site-packages/flask_restful/__init__.py", line 583, in dispatch_request
    resp = meth(*args, **kwargs)
  File "/home/farid/lab/Restfulness-core-api/venv/lib/python3.6/site-packages/flask_jwt_extended/view_decorators.py", line 108, in wrapper
    return fn(*args, **kwargs)
  File "/home/farid/lab/Restfulness-core-api/venv/lib/python3.6/site-packages/flasgger/utils.py", line 273, in wrapper
    return function(*args, **kwargs)
  File "/home/farid/lab/Restfulness-core-api/resources/Links.py", line 24, in get
    bookmarked_links_names.append(link.get_address_name())
AttributeError: 'str' object has no attribute 'get_address_name'

```

Also when you are adding a link it should return the added link id, not the whole structure which we already sent!
At the end the reason you didn't see the first error is you didn't write proper tests for the important parts, please write more tests, using a test coverage library could be helpful like this one:
https://coverage.readthedocs.io/en/coverage-5.3/

pesarkhobeee commented:
Dear @farbodahm 
I was testing this PR in action, I noticed your are returning the whole data structure which we already sent in the post -> user/links, it should return only the added new link's id
And in the get -> user/links you are returning only the links id! it should return the whole list
Also it should support user/links/[id] to get a specific link information.
Recheck your knowledge about RESTful design by reading articles like this:
https://stackoverflow.com/questions/671118/what-exactly-is-restful-programming

Also please use a test coverage library in order to be confident that we are writing enough tests:
https://coverage.readthedocs.io/en/coverage-5.3/

farbodahm committed:
Return only ID of new created link

farbodahm committed:
Return whole information about links

farbodahm committed:
Get specified link id

farbodahm committed:
Get id for deleting link from URL

farbodahm committed:
Changed routes addresses & seprated links related modules

farbodahm committed:
Documentation: Updated doc based on new changes

farbodahm committed:
Tests: Added more tests

farbodahm commented:
Dear @pesarkhobeee ,
I just added new commits to PR which I fixed the problems you mentioned before and added some new test;
Also I separated  `Links` related routes from `User` related routes to have better documentation.

pesarkhobeee commented:
@farbodahm well done :+1:
